### PR TITLE
bfloat16: vectorized unary ops

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35172 bfloat16: enable basic math function
* **#35092 bfloat16: vectorized unary ops**

Differential Revision: [D20721147](https://our.internmc.facebook.com/intern/diff/D20721147)

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
@uyongw, @Jianhui-Li, @jgong5, @hongzhen1

**Comment by XiaobingSuper** [comment]
@uyongw, @Jianhui-Li, @jgong5, @hongzhen1

File: test/test_torch.py [hunk 1]
```diff
@@ -14293,15 +14347,23 @@ def test_hardswish(self, device, dtype):
         self.assertEqual(inputTensorCpy, expectedOutputTensor, precision_4dps)
 
     @onlyCPU
-    @dtypes(torch.float, torch.double)
+    @dtypes(torch.float, torch.double, torch.bfloat16)
```
Comment (ngimel) [inline test/test_torch.py]: you can use @precisionOverride decorators here instead of manually setting precision

File: test/test_torch.py [hunk 2]
```diff
@@ -6097,6 +6140,8 @@ def test_pow(self, device):
     def test_neg(self, device):
         int_types = [torch.int, torch.short, torch.int8, torch.uint8]
         float_types = [torch.float, torch.double, torch.long]
+        if device == 'cpu':
```
Comment (ngimel) [inline test/test_torch.py]: `if self.device_type == "cpu"`

**Review by ngimel — CHANGES_REQUESTED** [review]
Please do the tests as described in #35082

File: test/test_torch.py [hunk 3]
```diff
@@ -16614,10 +16615,12 @@ def inner(self, device, dtype):
     ('erfc', '', _small_3d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
     ('exp', '', _small_3d, lambda t, d: [], 1e-2, 1e-5, 1e-5, _float_types),
     ('expm1', '', _small_3d, lambda t, d: [], 1e-2, 1e-5, 1e-5, _float_types),
-    ('reciprocal', '', _small_3d, lambda t, d: [], 1e-1, 1e-5, 1e-5, _float_types),
-    ('floor', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types),
-    ('frac', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types),
-    ('neg', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types2),
+    ('reciprocal', '', _small_3d, lambda t, d: [], 1e-1, 1e-2, 1e-5, _float_types),
+    ('reciprocal', 'no_zero', lambda t, d: _small_3d(t, d, has_zeros=False),
```
Comment (ngimel) [inline test/test_torch.py]: Why do you need a special test for no_zero? `nan` results are correctly handled in comparisons.

Comment (XiaobingSuper) [reply test/test_torch.py]: if we have zero number in input tensor, the output will have **inf** value for float32 and bfloat16, and then get a **nan** value(inf -inf), which report a error **nan is not less than 1e-2**.

Comment (ngimel) [reply test/test_torch.py]: You can change the line `self.assertEqual(cpu_result, device_result, prec=precision, exact_dtype=False)` to 
```
self.assertEqual(cpu_result, device_result, prec=precision, exact_dtype=False, allow_inf=True)
```
in the test itself to handle this.

File: test/test_torch.py [hunk 4]
```diff
@@ -16401,6 +16401,7 @@ def inner(self, device, dtype):
     ('clamp_max', '', _medium_2d, lambda t, d: [1], 1e-2, 1e-2, 1e-5, _types, [torch.bfloat16]),
     ('clone', '', _medium_2d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
     ('contiguous', '', _medium_2d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
+    ('conj', '', _small_3d, lambda t, d: [], 1e-5, 1e-2, 1e-5, _types, [torch.bfloat16], False),
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_torch.py]: changed.

**Review by ngimel — APPROVED** [review]
Small comments re: precision, also, can you please add a comment to the lines above tensor_op_tests, saying that argument after dtype_list is types to be tested on cpu?
After this line
```
# - dtype_list (=_types), a list of torch dtypes to test the op(s) with
```
add
```
# - cpu_dtype_list (=[]) ...
```

File: test/test_torch.py [hunk 5]
```diff
@@ -16627,10 +16634,10 @@ def inner(self, device, dtype):
     ('erfc', '', _small_3d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _float_types),
     ('exp', '', _small_3d, lambda t, d: [], 1e-2, 1e-5, 1e-5, _float_types),
     ('expm1', '', _small_3d, lambda t, d: [], 1e-2, 1e-5, 1e-5, _float_types),
-    ('reciprocal', '', _small_3d, lambda t, d: [], 1e-1, 1e-5, 1e-5, _float_types),
-    ('floor', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types),
-    ('frac', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types),
-    ('neg', '', _small_3d, lambda t, d: [], 1e-5, 1e-5, 1e-5, _float_types2),
+    ('reciprocal', '', _small_3d, lambda t, d: [], 1e-1, 1e-2, 1e-5, _float_types, [torch.bfloat16]),
+    ('floor', '', _small_3d, lambda t, d: [], 1e-5, 1e-2, 1e-5, _float_types),
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply test/test_torch.py]: changed.

**Comment by XiaobingSuper** [comment]
@ngimel , thanks for your review.

**Comment by ngimel** [comment]
ROCm test failure is legit, even though it's not entirely your fault. In `_make_tensors` we have 
```
    if dtype not in _float_types2:
        t = torch.randint(0, 10, shape, device=device)
        return t.to(_convert_t(dtype, device))
```
which will fill bfloat16 tensors with integer values in regular cpu runs, but won't do it (and instead fall down to fill them with randn) if the same test is run on an amd machine. This in turn will produce higher error that we are seeing. 
That way of filling tensors is wrong, you should fix it to 
```
    if not dtype.is_floating_point: 
        t = torch.randint(0, 10, shape, device=device)
        return t.to(_convert_t(dtype, device))
```
and raise the tolerance for the failing test. Sorry about that.

**Comment by ngimel** [comment]
CI failures are real, you need to increase tolerance of reciprocal test.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/cpu/vec256/vec256_bfloat16.h [hunk 6]
```diff
@@ -186,13 +186,109 @@ template <> class Vec256<BFloat16> {
...
+    auto ones = _mm256_set1_ps(1);
+    auto o1 = _mm256_div_ps(ones, _mm256_sqrt_ps(lo));
+    auto o2 = _mm256_div_ps(ones, _mm256_sqrt_ps(hi));
+    return cvtfp32_bf16(o1, o2);
+  }
+  // Comparison using the _CMP_**_OQ predicate.
+  //   `O`: get false if an operand is NaN
+  //   `Q`: do not raise if an operand is NaN
+  template<int imm8>
+  Vec256<BFloat16> compare(const Vec256<BFloat16>& other) const {
```
Comment (xuhdev) [inline aten/src/ATen/cpu/vec256/vec256_bfloat16.h:265]: Would you mind letting #35844 and #36106 implement the comparison operators? The reason is that those two PRs reduce code redundancy by using the same block of code across comparison, bitwise, and arithmetic operators. This function here virtually repeated the code snippet that implements other binary operators. Thanks!

Comment (XiaobingSuper) [reply aten/src/ATen/cpu/vec256/vec256_bfloat16.h:265]: @xuhdev , there has no conflict with them,  we can apply your PRs after @ngimel merge this PR. Thanks!

**Pull request closed without merging**